app/commands/generate_missing_mappings_command.rb

```diff
@@ -0,0 +1,318 @@
+  end
+
+  def load_embedding_data(output_taxonomy)
+    embeddings = nil
+    spinner("Loading embeddings for #{output_taxonomy}") do
+      files = Dir.glob(File.join("data/integrations/#{output_taxonomy}/embeddings", "_*.txt"))
+      embeddings = files.each_with_object({}) do |partition, embedding_data|
+        File.foreach(partition) do |line|
+          word, vector_str = line.chomp.split(":", 2)
+          vector = vector_str.split(", ").map { |num| BigDecimal(num).to_f }
+          embedding_data[word] = vector
+        end
+      end
+    end
+    embeddings
+  end
+
+  def index_embedding_data(embedding_data:, index_name:)
+    spinner("Indexing embeddings for #{index_name}") do
+      @qdrant_client.collections.delete(collection_name: index_name)
+      @qdrant_client.collections.create(
+        collection_name: index_name,
+        vectors: { size: 1536, distance: "Cosine" },
+      )
+
+      points = embedding_data.map.with_index do |(key, value), index|
+        {
+          id: index + 1,
+          vector: value,
+          payload: { index_name => key },
+        }
+      end
+
+      points.each_slice(100) do |batch|
+        @qdrant_client.points.upsert(
+          collection_name: index_name,
+          points: batch,
+        )
+      end
+    end
+  end
+
+  def generate_and_evaluate_mappings_for_group(
+    missing_mapping_group:,
+    index_name:
+  )
+    spinner("Generating and evaluating mappings for each Shopify category") do
+      destination_taxonomy_ids_by_full_name = load_destination_taxonomy_ids(missing_mapping_group[:output_taxonomy])
+      mapping_file_path = "data/integrations/#{missing_mapping_group[:output_taxonomy]}/mappings/from_shopify.yml"
+      mapping_data = YAML.load_file(mapping_file_path)
+
+      disagree_messages = []
+      missing_mapping_group[:category_ids_full_names].each do |source_category_id, source_category_name|
+        generated_mapping = generate_mapping(
+          source_category_id:,
+          source_category_name:,
+          index_name:,
+          destination_taxonomy_ids_by_full_name:,
+        )
+        mapping_data["rules"] << generated_mapping[:new_entry]
+        disagree_messages << generated_mapping[:mapping_to_be_graded] if generated_mapping[:grading_result] == "No"
+      end
+
+      mapping_data["rules"].sort_by! { |rule| rule["input"]["product_category_id"] }
+      File.write(mapping_file_path, mapping_data.to_yaml)
+
+      write_disagree_messages(disagree_messages) if disagree_messages.any?
+    end
+  end
+
+  def load_destination_taxonomy_ids(output_taxonomy)
+    logger.debug("Loading destination taxonomy IDs for #{output_taxonomy}")
+    YAML.load_file("data/integrations/#{output_taxonomy}/full_names.yml").each_with_object({}) do |category, hash|
+      hash[category["full_name"]] = category["id"]
+    end
+  end
+
+  def generate_missing_mappings_for_groups(missing_mapping_groups)
+    missing_mapping_groups.each do |missing_mapping_group|
```

Comment:
I'd suggest a frame around this each as well

Reply:
It is one line 135 `frame("Generating mappings for #{input_taxonomy} -> #{output_taxonomy}") do`

Reply:
Yeah I was thinking an outer frame? Or is in practice this loop size==1?